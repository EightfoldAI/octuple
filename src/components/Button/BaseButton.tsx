import React, { FC, Ref } from 'react';
import {
    ButtonIconAlign,
    ButtonShape,
    ButtonSize,
    ButtonTextAlign,
    ButtonTheme,
    ButtonWidth,
    InternalButtonProps,
    SplitButton,
} from './';
import { Icon, IconSize } from '../Icon';
import { Badge } from '../Badge';
import { Breakpoints, useMatchMedia } from '../../hooks/useMatchMedia';
import { mergeClasses } from '../../shared/utilities';

import styles from './button.module.scss';

export const BaseButton: FC<InternalButtonProps> = React.forwardRef(
    (
        {
            alignIcon = ButtonIconAlign.Left,
            alignText = ButtonTextAlign.Center,
            allowDisabledFocus = false,
            ariaLabel,
            buttonWidth = ButtonWidth.fitContent,
            checked = false,
            classNames,
            counter,
            disabled = false,
            disruptive = false,
            dropShadow = false,
            htmlType,
            iconProps,
            id,
            onClick,
            onContextMenu,
            shape = ButtonShape.Pill,
            size = ButtonSize.Flex,
            split,
            splitButtonChecked = false,
            splitButtonProps,
            style,
            text,
            theme,
            toggle,
            type,
            ...rest
        },
        ref: Ref<HTMLButtonElement>
    ) => {
        const largeScreenActive: boolean = useMatchMedia(Breakpoints.Large);
        const mediumScreenActive: boolean = useMatchMedia(Breakpoints.Medium);
        const smallScreenActive: boolean = useMatchMedia(Breakpoints.Small);
        const xSmallScreenActive: boolean = useMatchMedia(Breakpoints.XSmall);

        const counterExists: boolean = !!counter;
        const iconExists: boolean = !!iconProps;
        const textExists: boolean = !!text;

        const buttonBaseSharedClassNames: string = mergeClasses([
            classNames,
            {
                [styles.buttonSize3]:
                    size === ButtonSize.Flex && largeScreenActive,
            },
            {
                [styles.buttonSize2]:
                    size === ButtonSize.Flex && mediumScreenActive,
            },
            {
                [styles.buttonSize2]:
                    size === ButtonSize.Flex && smallScreenActive,
            },
            {
                [styles.buttonSize1]:
                    size === ButtonSize.Flex && xSmallScreenActive,
            },
            { [styles.buttonSize1]: size === ButtonSize.Large },
            { [styles.buttonSize2]: size === ButtonSize.Medium },
            { [styles.buttonSize3]: size === ButtonSize.Small },
            { [styles.pillShape]: shape === ButtonShape.Pill },
            { [styles.dropShadow]: dropShadow },
            { [styles.dark]: theme === ButtonTheme.dark },
        ]);

        const buttonBaseClassNames: string = mergeClasses([
            buttonBaseSharedClassNames,
            { [styles.buttonStretch]: buttonWidth === ButtonWidth.fill },
            { [styles.splitLeft]: split },
            { [styles.left]: alignText === ButtonTextAlign.Left },
            { [styles.right]: alignText === ButtonTextAlign.Right },
            {
                [styles.iconLeft]:
                    iconExists && alignIcon === ButtonIconAlign.Left,
            },
            {
                [styles.iconRight]:
                    iconExists && alignIcon === ButtonIconAlign.Right,
            },
            { [styles.disabled]: allowDisabledFocus || disabled },
        ]);

        const buttonTextClassNames: string = mergeClasses([
            {
                [styles.buttonText3]:
                    size === ButtonSize.Flex && largeScreenActive,
            },
            {
                [styles.buttonText2]:
                    size === ButtonSize.Flex && mediumScreenActive,
            },
            {
                [styles.buttonText2]:
                    size === ButtonSize.Flex && smallScreenActive,
            },
            {
                [styles.buttonText1]:
                    size === ButtonSize.Flex && xSmallScreenActive,
            },
            { [styles.buttonText1]: size === ButtonSize.Large },
            { [styles.buttonText2]: size === ButtonSize.Medium },
            { [styles.buttonText3]: size === ButtonSize.Small },
        ]);

        const badgeClassNames: string = mergeClasses([
            styles.counter,
            buttonTextClassNames,
        ]);

        const getButtonIconSize = (): IconSize => {
            let iconSize: IconSize;
            if (size === ButtonSize.Flex && largeScreenActive) {
                iconSize = IconSize.Small;
            } else if (
                size === ButtonSize.Flex &&
                (mediumScreenActive || smallScreenActive)
            ) {
                iconSize = IconSize.Medium;
            } else if (size === ButtonSize.Flex && xSmallScreenActive) {
                iconSize = IconSize.Large;
            } else if (size === ButtonSize.Large) {
                iconSize = IconSize.Large;
            } else if (size === ButtonSize.Medium) {
                iconSize = IconSize.Medium;
            } else if (size === ButtonSize.Small) {
                iconSize = IconSize.Small;
            }
            return iconSize;
        };

        const getButtonIcon = (): JSX.Element => (
            <Icon
                {...iconProps}
                classNames={mergeClasses([styles.icon, iconProps.classNames])}
                size={getButtonIconSize()}
            />
        );

        const getButtonContent = (
            buttonTextClassNames: string,
            text: string
        ): JSX.Element => (
            <span className={buttonTextClassNames}>
                {text ? text : 'Button'}
                {counterExists && (
                    <Badge classNames={badgeClassNames}>{counter}</Badge>
                )}
            </span>
        );

        return (
            <>
                <button
                    ref={ref}
                    {...rest}
                    aria-checked={toggle ? !!checked : undefined}
                    aria-disabled={allowDisabledFocus}
                    aria-label={ariaLabel}
                    aria-pressed={toggle ? !!checked : undefined}
                    defaultChecked={checked}
                    disabled={disabled}
                    className={buttonBaseClassNames}
                    id={id}
                    onClick={!allowDisabledFocus ? onClick : null}
                    style={style}
                    type={htmlType}
                >
                    {iconExists && !textExists && getButtonIcon()}
                    {counterExists && !textExists && (
                        <Badge classNames={badgeClassNames}>{counter}</Badge>
                    )}
                    {iconExists && textExists && (
                        <span>
                            {getButtonIcon()}
                            {getButtonContent(buttonTextClassNames, text)}
                        </span>
                    )}
                    {!iconExists &&
                        getButtonContent(buttonTextClassNames, text)}
                </button>
                {split && (
                    <SplitButton
                        {...splitButtonProps}
                        classNames={
                            buttonBaseSharedClassNames +
                            ' ' +
                            splitButtonProps?.classNames
                        }
                        checked={splitButtonChecked}
                        disruptive={disruptive}
                        dropShadow={dropShadow}
                        onClick={
                            !splitButtonProps?.allowDisabledFocus
                                ? onContextMenu
                                : null
                        }
                        shape={shape}
                        size={size}
                        split={split}
                        theme={theme}
                        type={type}
                    />
                )}
            </>
        );
    }
);
