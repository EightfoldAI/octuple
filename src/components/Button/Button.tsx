import React, { FC, Ref, useContext, useEffect, useRef, useState } from 'react';
import DisabledContext, { Disabled } from '../ConfigProvider/DisabledContext';
import { SizeContext, Size } from '../ConfigProvider';
import {
  ButtonIconAlign,
  ButtonProps,
  ButtonShape,
  ButtonSize,
  ButtonTextAlign,
  ButtonVariant,
  ButtonWidth,
  SplitButton,
} from '.';
import { Icon, IconSize } from '../Icon';
import { Badge } from '../Badge';
import { InnerNudge, NudgeAnimation, NudgeProps } from './Nudge';
import { Breakpoints, useMatchMedia } from '../../hooks/useMatchMedia';
import { mergeClasses } from '../../shared/utilities';
import { Loader, LoaderSize } from '../Loader';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { useMergedRefs } from '../../hooks/useMergedRefs';
import { useNudge } from './Nudge/Hooks/useNudge';

import styles from './button.module.scss';

export const Button: FC<ButtonProps> = React.forwardRef(
  (props: ButtonProps, ref: Ref<HTMLButtonElement>) => {
    const {
      alignIcon = ButtonIconAlign.Left,
      alignText = ButtonTextAlign.Center,
      allowDisabledFocus = false,
      ariaLabel,
      buttonWidth = ButtonWidth.fitContent,
      checked = false,
      classNames,
      configContextProps = {
        noDisabledContext: false,
        noSizeContext: false,
      },
      counter,
      disabled = false,
      disruptive = false,
      dropShadow = false,
      floatingButtonProps,
      nudgeProps: defaultNudgeProps,
      htmlType,
      iconProps,
      id,
      loading = false,
      onClick,
      onContextMenu,
      prefixIconProps,
      shape = ButtonShape.Pill,
      size = ButtonSize.Medium,
      split,
      splitButtonChecked = false,
      splitButtonProps,
      style,
      text,
      toggle,
      transparent = false,
      type, // TODO: Remove in Octuple v3.0.0 and use `variant` only.
      variant = ButtonVariant.Default,
      ...rest
    } = props;
    const largeScreenActive: boolean = useMatchMedia(Breakpoints.Large);
    const mediumScreenActive: boolean = useMatchMedia(Breakpoints.Medium);
    const smallScreenActive: boolean = useMatchMedia(Breakpoints.Small);
    const xSmallScreenActive: boolean = useMatchMedia(Breakpoints.XSmall);

    const internalRef: React.MutableRefObject<HTMLButtonElement> =
      useRef<HTMLButtonElement>(null);

    const mergedRef = useMergedRefs(internalRef, ref);

    const htmlDir: string = useCanvasDirection();

    const contextuallyDisabled: Disabled = useContext(DisabledContext);
    const mergedDisabled: boolean = configContextProps.noDisabledContext
      ? disabled
      : contextuallyDisabled || disabled;

    const contextuallySized: Size = useContext(SizeContext);
    const mergedSize = configContextProps.noSizeContext
      ? size
      : contextuallySized || size;

    const counterExists: boolean = !!counter;
    const iconExists: boolean = !!iconProps;
    const prefixIconExists: boolean = !!prefixIconProps;
    const textExists: boolean = !!text;

    const [nudgeProps, setNudgeProps] = useState<NudgeProps>(defaultNudgeProps);
    const innerNudgeRef: React.MutableRefObject<HTMLSpanElement> =
      useRef<HTMLSpanElement>(null);

    useEffect(() => {
      setNudgeProps(
        props.nudgeProps
          ? props.nudgeProps
          : {
              animation: NudgeAnimation.Background,
              delay: 2000,
              iterations: 1,
              enabled: false,
            }
      );
    }, [nudgeProps?.enabled]);

    useNudge(disruptive, nudgeProps, [internalRef, innerNudgeRef], styles);

    // TODO: Remove in Octuple v3.0.0 and use `variant` only.
    // For now, if `type` has a value use it, else use `variant`.
    const mergedVariant: ButtonVariant = !!type ? type : variant;

    const buttonSharedClassNames: string = mergeClasses([
      classNames,
      styles.button,
      { [styles.buttonDefault]: mergedVariant === ButtonVariant.Default },
      { [styles.buttonNeutral]: mergedVariant === ButtonVariant.Neutral },
      { [styles.buttonPrimary]: mergedVariant === ButtonVariant.Primary },
      { [styles.buttonSecondary]: mergedVariant === ButtonVariant.Secondary },
      { [styles.buttonSystemUi]: mergedVariant === ButtonVariant.SystemUI },
      {
        [styles.buttonDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Default,
      },
      {
        [styles.buttonPrimaryDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Primary,
      },
      {
        [styles.buttonSecondaryDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Secondary,
      },
      {
        [styles.transparent]:
          transparent && mergedVariant === ButtonVariant.SystemUI,
      },
      {
        [styles.buttonSmall]:
          mergedSize === ButtonSize.Flex && largeScreenActive,
      },
      {
        [styles.buttonMedium]:
          mergedSize === ButtonSize.Flex && mediumScreenActive,
      },
      {
        [styles.buttonMedium]:
          mergedSize === ButtonSize.Flex && smallScreenActive,
      },
      {
        [styles.buttonLarge]:
          mergedSize === ButtonSize.Flex && xSmallScreenActive,
      },
      { [styles.buttonLarge]: mergedSize === ButtonSize.Large },
      { [styles.buttonMedium]: mergedSize === ButtonSize.Medium },
      { [styles.buttonSmall]: mergedSize === ButtonSize.Small },
      { [styles.pillShape]: shape === ButtonShape.Pill },
      {
        [styles.roundShape]:
          shape === ButtonShape.Round && !split && !textExists,
      },
      { [styles.dropShadow]: dropShadow },
      { [styles.disabled]: allowDisabledFocus || mergedDisabled },
      { [styles.floating]: floatingButtonProps?.enabled },
      { [styles.buttonRtl]: htmlDir === 'rtl' },
      {
        [styles.buttonConic]:
          !disruptive &&
          nudgeProps?.enabled &&
          nudgeProps?.animation === NudgeAnimation.Conic,
      },
    ]);

    const buttonClassNames: string = mergeClasses([
      buttonSharedClassNames,
      { [styles.buttonStretch]: buttonWidth === ButtonWidth.fill },
      { [styles.splitLeft]: split },
      { [styles.left]: alignText === ButtonTextAlign.Left },
      { [styles.right]: alignText === ButtonTextAlign.Right },
      {
        [styles.iconLeft]: iconExists && alignIcon === ButtonIconAlign.Left,
      },
      {
        [styles.iconRight]: iconExists && alignIcon === ButtonIconAlign.Right,
      },
      { [styles.loading]: loading },
    ]);

    const buttonTextClassNames: string = mergeClasses([
      {
        [styles.buttonTextSmall]:
          mergedSize === ButtonSize.Flex && largeScreenActive,
      },
      {
        [styles.buttonTextMedium]:
          mergedSize === ButtonSize.Flex && mediumScreenActive,
      },
      {
        [styles.buttonTextMedium]:
          mergedSize === ButtonSize.Flex && smallScreenActive,
      },
      {
        [styles.buttonTextLarge]:
          mergedSize === ButtonSize.Flex && xSmallScreenActive,
      },
      { [styles.buttonTextLarge]: mergedSize === ButtonSize.Large },
      { [styles.buttonTextMedium]: mergedSize === ButtonSize.Medium },
      { [styles.buttonTextSmall]: mergedSize === ButtonSize.Small },
    ]);

    const badgeClassNames: string = mergeClasses([
      styles.counter,
      buttonTextClassNames,
    ]);

    const getButtonIconSize = (): IconSize => {
      let iconSize: IconSize;
      if (mergedSize === ButtonSize.Flex && largeScreenActive) {
        iconSize = IconSize.Small;
      } else if (
        mergedSize === ButtonSize.Flex &&
        (mediumScreenActive || smallScreenActive)
      ) {
        iconSize = IconSize.Medium;
      } else if (mergedSize === ButtonSize.Flex && xSmallScreenActive) {
        iconSize = IconSize.Large;
      } else if (mergedSize === ButtonSize.Large) {
        iconSize = IconSize.Large;
      } else if (mergedSize === ButtonSize.Medium) {
        iconSize = IconSize.Medium;
      } else if (mergedSize === ButtonSize.Small) {
        iconSize = IconSize.Small;
      }
      return iconSize;
    };

    const getLoaderSize = (): LoaderSize => {
      let loaderSize: LoaderSize;
      if (size === ButtonSize.Flex && largeScreenActive) {
        loaderSize = LoaderSize.Small;
      } else if (
        size === ButtonSize.Flex &&
        (mediumScreenActive || smallScreenActive)
      ) {
        loaderSize = LoaderSize.Medium;
      } else if (size === ButtonSize.Flex && xSmallScreenActive) {
        loaderSize = LoaderSize.Large;
      } else if (size === ButtonSize.Large) {
        loaderSize = LoaderSize.Large;
      } else if (size === ButtonSize.Medium) {
        loaderSize = LoaderSize.Medium;
      } else if (size === ButtonSize.Small) {
        loaderSize = LoaderSize.Small;
      }
      return loaderSize;
    };

    const getButtonLoader = (): JSX.Element =>
      loading && (
        <Loader
          classNames={styles.loader}
          dotClassNames={styles.loaderDot}
          size={getLoaderSize()}
        />
      );

    const getButtonIcon = (): JSX.Element => (
      <Icon
        {...iconProps}
        classNames={mergeClasses([styles.icon, iconProps.classNames])}
        size={getButtonIconSize()}
      />
    );

    const getPrefixIcon = (): JSX.Element => (
      <Icon
        {...prefixIconProps}
        classNames={mergeClasses([
          styles.icon,
          styles.prefixIcon,
          prefixIconProps.classNames,
        ])}
        size={getButtonIconSize()}
      />
    );

    const getButtonContent = (
      buttonTextClassNames: string,
      text: string
    ): JSX.Element => (
      <span className={buttonTextClassNames}>
        {text ? text : 'Button'}
        {counterExists && <Badge classNames={badgeClassNames}>{counter}</Badge>}
      </span>
    );

    return (
      <>
        <button
          {...rest}
          ref={mergedRef}
          aria-checked={toggle ? !!checked : undefined}
          aria-disabled={mergedDisabled || loading}
          aria-label={ariaLabel}
          aria-pressed={toggle ? !!checked : undefined}
          defaultChecked={checked}
          disabled={(!allowDisabledFocus && mergedDisabled) || loading}
          className={buttonClassNames}
          id={id}
          onClick={!allowDisabledFocus ? onClick : null}
          style={style}
          type={htmlType}
        >
          <InnerNudge
            classNames={mergeClasses([
              styles.innerNudge,
              {
                [styles.conic]: nudgeProps?.animation === NudgeAnimation.Conic,
              },
            ])}
            disruptive={disruptive}
            id={id ? `${id}-nudge` : 'button-nudge'}
            nudgeProps={nudgeProps}
            ref={innerNudgeRef}
            style={style}
          />
          {iconExists && prefixIconExists && !textExists && (
            <span>
              {getButtonIcon()}
              {getPrefixIcon()}
            </span>
          )}
          {iconExists &&
            !textExists &&
            !prefixIconExists &&
            !counterExists &&
            getButtonIcon()}
          {counterExists && !textExists && !loading && !iconExists && (
            <Badge classNames={badgeClassNames}>{counter}</Badge>
          )}
          {iconExists && counterExists && !textExists && !loading && (
            <span>
              {getButtonIcon()}
              <Badge classNames={badgeClassNames}>{counter}</Badge>
              {prefixIconExists && getPrefixIcon()}
            </span>
          )}
          {iconExists && textExists && (
            <span>
              {getButtonIcon()}
              {getButtonContent(buttonTextClassNames, text)}
              {prefixIconExists && getPrefixIcon()}
            </span>
          )}
          {!iconExists && getButtonContent(buttonTextClassNames, text)}
          {getButtonLoader()}
        </button>
        {split && (
          <SplitButton
            {...splitButtonProps}
            classNames={
              buttonSharedClassNames + ' ' + splitButtonProps?.classNames
            }
            checked={splitButtonChecked}
            disruptive={disruptive}
            dropShadow={dropShadow}
            nudgeProps={nudgeProps}
            onClick={
              !splitButtonProps?.allowDisabledFocus ? onContextMenu : null
            }
            shape={shape}
            size={mergedSize}
            split={split}
            variant={mergedVariant}
          />
        )}
      </>
    );
  }
);
