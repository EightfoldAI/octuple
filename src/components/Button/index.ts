export * from './Button.types';
export * from './BaseButton';
export * from './DefaultButton/DefaultButton';
export * from './NeutralButton/NeutralButton';
export * from './PrimaryButton/PrimaryButton';
export * from './SecondaryButton/SecondaryButton';
export * from './SplitButton/SplitButton';
export * from './TwoStateButton/TwoStateButton';
