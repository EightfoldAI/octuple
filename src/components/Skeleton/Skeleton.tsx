'use client';

import React, { FC, Ref } from 'react';
import {
  SkeletonAnimation,
  SkeletonProps,
  SkeletonVariant,
} from './Skeleton.types';
import { ConditionalWrapper, mergeClasses } from '../../shared/utilities';

import styles from './skeleton.module.scss';

const IGNORE_CUSTOM_HEIGHT_VARIANTS = [
  SkeletonVariant.ButtonSmall,
  SkeletonVariant.ButtonMedium,
  SkeletonVariant.ButtonLarge,
];

export const Skeleton: FC<SkeletonProps> = React.forwardRef(
  (
    {
      classNames,
      style,
      width,
      height,
      fullWidth,
      variant = SkeletonVariant.Rectangular,
      animation = SkeletonAnimation.Wave,
      animating = true,
      children,
      'data-test-id': dataTestId,
    },
    ref: Ref<HTMLDivElement>
  ) => {
    const skeletonClasses: string = mergeClasses([
      classNames,
      styles.skeleton,
      {
        [styles.pulse]: animation === SkeletonAnimation.Pulse && animating,
        [styles.wave]: animation === SkeletonAnimation.Wave && animating,
        [styles.circular]: variant === SkeletonVariant.Circular,
        [styles.pill]: variant === SkeletonVariant.Pill,
        [styles.rounded]: variant === SkeletonVariant.Rounded,
        [styles.fullWidth]: fullWidth,
        [styles.buttonSmall]: variant === SkeletonVariant.ButtonSmall,
        [styles.buttonMedium]: variant === SkeletonVariant.ButtonMedium,
        [styles.buttonLarge]: variant === SkeletonVariant.ButtonLarge,
        [styles.children]: children,
      },
    ]);

    const customStyle: React.CSSProperties = {
      ...(!fullWidth ? { width } : {}),
      ...(!IGNORE_CUSTOM_HEIGHT_VARIANTS.includes(variant) ? { height } : {}),
    };

    const getSkeleton = (): JSX.Element => (
      <div
        ref={ref}
        className={skeletonClasses}
        tabIndex={-1}
        data-test-id={dataTestId}
        style={{ ...style, ...customStyle }}
      />
    );

    return (
      <ConditionalWrapper
        condition={!!children}
        wrapper={(skeleton) => (
          <div className={styles.skeletonContainer}>
            {children}
            {animating && skeleton}
          </div>
        )}
      >
        {getSkeleton()}
      </ConditionalWrapper>
    );
  }
);
