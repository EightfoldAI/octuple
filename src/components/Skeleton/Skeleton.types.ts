import { OcBaseProps } from '../OcBase';

export enum SkeletonVariant {
    Circular = 'circular',
    Rectangular = 'rectangular',
    Pill = 'pill',
    Rounded = 'rounded',
    ButtonSmall = 'buttonSmall',
    ButtonMedium = 'buttonMedium',
    ButtonLarge = 'buttonLarge',
}

export enum SkeletonAnimation {
    Pulse = 'pulse',
    Wave = 'wave',
}

type Dimension = number | string;

export interface SkeletonProps extends OcBaseProps<HTMLDivElement> {
    /**
     * Is animating or not
     * @default true
     */
    animating?: boolean;
    /**
     * Animation of the skeleton component
     * @default SkeletonAnimation.Wave
     */
    animation?: SkeletonAnimation;
    /**
     * If the skeleton is full width, will ignore the width prop
     */
    fullWidth?: boolean;
    /**
     * Height of the skeleton
     */
    height?: Dimension;
    /**
     * Visible variant of the skeleton
     * @default SkeletonVariant.Rectangular
     */
    variant?: SkeletonVariant;
    /**
     * Width of the skeleton
     */
    width?: Dimension;
}
