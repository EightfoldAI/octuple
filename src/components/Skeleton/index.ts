export * from './Skeleton';
export * from './Skeleton.types';
