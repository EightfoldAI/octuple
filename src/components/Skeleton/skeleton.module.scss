$skeleton-pulse-animation: pulse 1.5s ease-in-out 0.5s infinite;
$skeleton-wave-animation: wave 1.6s linear 0.5s infinite;
$skeleton-wave-background: linear-gradient(
    90deg,
    transparent,
    #e2e6ea,
    transparent
);

.skeleton-container {
    position: relative;
    width: fit-content;
    overflow: hidden;
}

.skeleton {
    background: var(--grey-color-20);

    &.pulse {
        animation: $skeleton-pulse-animation;
    }

    &.wave {
        position: relative;
        overflow: hidden;

        &:after {
            animation: $skeleton-wave-animation;
            background: $skeleton-wave-background;
            content: '';
            position: absolute;
            transform: translateX(-100%);
            bottom: 0;
            left: 0;
            right: 0;
            top: 0;
        }
    }

    &.children {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
    }

    &.pill {
        border-radius: $corner-radius-xl;
    }

    &.button-small {
        border-radius: $corner-radius-xl;
        height: 28px;
    }

    &.button-medium {
        border-radius: $corner-radius-xl;
        height: 36px;
    }

    &.button-large {
        border-radius: $corner-radius-xl;
        height: 44px;
    }

    &.rounded {
        border-radius: $corner-radius-m;
    }

    &.circular {
        border-radius: 50%;
    }

    &.full-width {
        width: 100%;
    }
}

@keyframes pulse {
    0% {
        opacity: 1;
    }

    50% {
        opacity: 0.4;
    }

    100% {
        opacity: 1;
    }
}

@keyframes wave {
    0% {
        transform: translateX(-100%);
    }

    50% {
        transform: translateX(100%);
    }

    100% {
        transform: translateX(100%);
    }
}
