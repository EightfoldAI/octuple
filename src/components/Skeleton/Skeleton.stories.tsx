import React from 'react';
import { Stories } from '@storybook/addon-docs';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { Skeleton } from './Skeleton';
import {
  SkeletonAnimation,
  SkeletonProps,
  SkeletonVariant,
} from './Skeleton.types';
import { Stack } from '../Stack';
import { Button } from '../Button';

export default {
  title: 'Skeleton',
  parameters: {
    docs: {
      page: (): JSX.Element => (
        <main>
          <article>
            <section></section>
            <section>
              <Stories includePrimary title="" />
            </section>
          </article>
        </main>
      ),
    },
  },
  argTypes: {
    animation: {
      options: [SkeletonAnimation.Pulse, SkeletonAnimation.Wave],
      control: { type: 'inline-radio' },
    },
    variant: {
      options: [
        SkeletonVariant.Circular,
        SkeletonVariant.Rounded,
        SkeletonVariant.Pill,
        SkeletonVariant.Rectangular,
        SkeletonVariant.ButtonSmall,
        SkeletonVariant.ButtonMedium,
        SkeletonVariant.ButtonLarge,
      ],
      control: { type: 'select' },
    },
  },
} as ComponentMeta<typeof Skeleton>;

const Default_Story: ComponentStory<typeof Skeleton> = (args) => (
  <Skeleton {...args} />
);

const Child_Wrapper_Story: ComponentStory<typeof Skeleton> = (args) => (
  <Skeleton {...args}>
    <Button text={'Sample button'} />
  </Skeleton>
);

const Sample_Usage_Story: ComponentStory<typeof Skeleton> = (args) => (
  <Stack direction="vertical" flexGap="xs" style={{ width: 210 }}>
    <Skeleton
      width={210}
      height={10}
      variant={SkeletonVariant.Rounded}
      {...args}
    />
    <Skeleton
      width={40}
      height={40}
      variant={SkeletonVariant.Circular}
      {...args}
    />
    <Skeleton width={210} height={60} {...args} />
    <Skeleton width={210} height={60} {...args} />
    <Stack justify="flex-end" gap="xs" fullWidth>
      <Skeleton width={60} variant={SkeletonVariant.ButtonSmall} {...args} />
      <Skeleton width={60} variant={SkeletonVariant.ButtonSmall} {...args} />
    </Stack>
  </Stack>
);

export const Default = Default_Story.bind({});
export const Child_Wrapper = Child_Wrapper_Story.bind({});
export const Sample_Usage = Sample_Usage_Story.bind({});

// Storybook 6.5 using Webpack >= 5.76.0 automatically alphabetizes exports,
// this line ensures they are exported in the desired order.
// See https://www.npmjs.com/package/babel-plugin-named-exports-order
export const __namedExportsOrder = ['Default', 'Child_Wrapper', 'Sample_Usage'];

const skeletonArgs: SkeletonProps = {
  animating: true,
  animation: SkeletonAnimation.Wave,
};

Default.args = {
  ...skeletonArgs,
  width: 80,
  height: 80,
  fullWidth: false,
  variant: SkeletonVariant.Rectangular,
};

Child_Wrapper.args = {
  ...skeletonArgs,
  variant: SkeletonVariant.Pill,
};

Sample_Usage.args = {
  ...skeletonArgs,
};
