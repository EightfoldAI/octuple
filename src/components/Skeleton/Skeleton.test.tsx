import React from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { Button } from '../Button';
import { Skeleton } from './Skeleton';
import {
  SkeletonAnimation,
  SkeletonProps,
  SkeletonVariant,
} from './Skeleton.types';
import { render } from '@testing-library/react';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia: any;

const skeletonProps: SkeletonProps = {
  animating: true,
  animation: SkeletonAnimation.Wave,
};

describe('Skeleton', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  test('Renders without crashing', () => {
    const { container } = render(
      <Skeleton {...skeletonProps} variant={SkeletonVariant.Pill}>
        <Button text={'Sample button'} />
      </Skeleton>
    );
    const skeleton = container.getElementsByClassName('skeleton');
    expect(() => container).not.toThrowError();
    expect(skeleton).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
