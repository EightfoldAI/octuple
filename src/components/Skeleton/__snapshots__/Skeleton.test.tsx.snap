// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton Renders without crashing 1`] = `
<div>
  <div
    class="skeleton-container"
  >
    <button
      aria-disabled="false"
      class="button button-default button-medium pill-shape"
    >
      <span
        class="button-text-medium"
      >
        Sample button
      </span>
    </button>
    <div
      class="skeleton wave children"
      tabindex="-1"
    />
  </div>
</div>
`;
